core/store/src/trie/shard_tries.rs

```diff
@@ -47,6 +47,15 @@ impl ShardTries {
         }
     }
 
+    pub fn add_new_shards(&mut self, shards: &Vec<ShardUId>) {
+        Arc::get_mut(&mut self.caches)
+            .unwrap()
```

Comment:
This `unwrap` feels very worrisome to me -- it isn't locally obvious that we guarantee that there's a single reference here. 

Looking at the broader context, the whole of 

```rust
#[derive(Clone)]
pub struct ShardTries {
    pub(crate) store: Arc<Store>,
    /// Cache reserved for client actor to use
    pub(crate) caches: Arc<HashMap<ShardUId, TrieCache>>,
    /// Cache for readers.
    pub(crate) view_caches: Arc<HashMap<ShardUId, TrieCache>>,
}
```

feels a bit suspect:

* why we arc each field separatelly instead of Arcing the whole thing once? 
* `Arc<Store>` doesn't make sense, because `Store` is already an `Arc`. 


:thinking: yeah, I think we should return to this bit after the PR with an eye towards refactring, for the present case, I think we can take advantage of the fact that the map is small and shallow (the `TrieCache` is another arc), and just re-build the map:

```rust
    pub fn add_new_shards(&mut self, shards: &[ShardUId]) {
        let add_empty_caches = |old_caches: &HashMap<ShardUId, TrieCache>| {
            old_caches
                .iter()
                .map(|(k, v)| (*k, v.clone()))
                .zip(shards.iter().map(|&k| (k, TrieCache::new())))
                .collect::<HashMap<_>>()
        };
        self.caches = add_empty_caches(&*self.caches);
        self.view_caches = add_empty_caches(&*self.caches);
    }
```

Reply:
That's a very good point. The `Arc` seems confusing. 

I actually moved this part of change to another PR https://github.com/near/nearcore/pull/4804 because I realized I need to add a RwLock to shard_tries, since now it can be modified. I'll fix your comments there.

Reply:
I checked the code and it seems that `Arc<Store>` is used in many other places in the code too. I don't have enough context to tell why that is the case, but I assume that is why `Arc<Store>` is used here. 

And the function `get_store` needs to return `Arc<Store>`, I guess that's why the field needs to be arc'ed individually

Reply:
might make sense to create an issue to get rid of `Arc<Store>` everywhere then -- seems like a minor technical debt. The main problem here is that `Arc`s and `RwLock`s are important large pieces of overall application architecture, so having fewer of those can make the code more understandable.